README.md
<div align="center">

<h1>React-tabllist</h1>

React-based list (table) components that support events, callback functions, and custom styles.

---

[![GitHub License](https://img.shields.io/github/license/oceanxy/react-tabllist.svg)](https://github.com/oceanxy/react-tabllist/blob/master/LICENSE)
[![NPM Version](https://img.shields.io/npm/v/react-tabllist.svg)](https://www.npmjs.com/package/react-tabllist)
[![Minified Size](https://img.shields.io/bundlephobia/min/react-tabllist.svg)](https://bundlephobia.com/result?p=react-tabllist)
[![CircleCI Status](https://img.shields.io/circleci/project/github/oceanxy/react-tabllist/master.svg)](https://circleci.com/gh/oceanxy/react-tabllist)
[![NPM Download](https://img.shields.io/npm/dw/react-tabllist.svg)](https://www.npmjs.com/package/react-tabllist)

[![Codecov](https://img.shields.io/codecov/c/github/oceanxy/react-tabllist/master.svg)](https://codecov.io/gh/oceanxy/react-tabllist)
[![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/oceanxy/react-tabllist.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/oceanxy/react-tabllist/context:javascript)
[![Dependency React](https://img.shields.io/npm/dependency-version/react-tabllist/peer/react.svg)]()
[![Last Commit](https://img.shields.io/github/last-commit/oceanxy/react-tabllist.svg)](https://github.com/oceanxy/react-tabllist)

[comment]: <> ([![Dependencies Status]&#40;https://david-dm.org/oceanxy/react-tabllist/status.svg&#41;]&#40;https://david-dm.org/oceanxy/react-tabllist&#41;&#41;)
[comment]: <> ([![DevDependencies Status]&#40;https://david-dm.org/oceanxy/react-tabllist/dev-status.svg&#41;]&#40;https://david-dm.org/oceanxy/react-tabllist?type=dev&#41;)
[comment]: <> ([![PeerDependencies Status]&#40;https://david-dm.org/oceanxy/react-tabllist/peer-status.svg&#41;]&#40;https://david-dm.org/oceanxy/react-tabllist?type=peer&#41;)
[comment]: <> ([![Gitter]&#40;https://img.shields.io/gitter/room/oceanxy/react-tabllist.svg&#41;]&#40;https://gitter.im/react-tabllist/community?utm_source=share-link&utm_medium=link&utm_campaign=share-link&#41;)

</div>

English | [简体中文](./README_zh-CN.md)

---

## 🔧 Install

```bash
npm install react-tabllist --save
```

## 📚 Usage

  ### Use in browser
  
  ```html
    <script src="https://cdn.bootcss.com/react/16.8.6/umd/react.development.js"></script>
    <script src="https://cdn.bootcss.com/react-dom/16.8.6/umd/react-dom.development.js"></script>
    <script src="https://cdn.bootcss.com/babel-standalone/7.0.0-beta.3/babel.min.js"></script>
    <script src='https://cdn.bootcss.com/lodash.js/4.17.11/lodash.core.js'></script>
    <script src='./react-tabllist.min.js'></script>
  
    <div id='example'></div>
    <script type='text/javascript'>
      var option = {
        data: [],
        property: {}
      }
  
      ReactDOM.render(
        React.createElement(ReactTabllist, option),
        document.getElementById('example')
      )
    </script>
  ```

  ### Use in ES6 & jsx
  
  ```jsx harmony
    import ReactTabllist from 'react-tabllist';
  
    // Use default configuration
    ReactDOM.render(<ReactTabllist />, mountNode);
  
    // Custom data and configuration items
    const props = {
      className: 'demo',
      property: {},
      data: [
        ['1st column', '2nd column', '3rd column'],
        ['1st cell', '2nd cell', '3rd cell']
      ]
    }
  
    ReactDOM.render(<ReactTabllist {...props} />, mountNode);
  
    /**
     * Event callback function
     * @param {SyntheticEvent} event
     * @param {Exposes} exposes An object containing exposed properties and methods
     */
    function handleScroll(event, exposes) {
      // some code
    }
  
    // Use React native events
    ReactDOM.render(<ReactTabllist {...props} onScroll={handleScroll} />, mountNode);
  
    // Use custom events built into components
    ReactDOM.render(<ReactTabllist {...props} custom_onScrollToEnd={handleScroll} />, mountNode);
  ```

## 💻 Development

  If you want to join the development to improve this component, please fork and start with the following command:

  ```bash
    $ git clone git@github.com:oceanxy/react-tabllist.git
    $ cd react-tabllist
    $ npm install
    $ npm start
  ```

  Open your browser and visit `http://localhost:3001`.

## 📝 Version Information

  [Change Log](./CHANGELOG.md)

## 🧰 Configuration

  |**props**                                    |description                                     |details                           |
  |---------------------------------------------|------------------------------------------------|----------------------------------|
  |**data** <br> `{[Array, Array, Array,...]}`  |Render table data                               |[props.data](#props.data)         |
  |**className** <br> `{string}`                |Table outermost container style sheet name      |''                                |
  |**property** <br> `{Object}`                 |Wrapper table properties                        |[props.property](#props.property) |

  ### props.className

  You can customize the style through the `props.classname` attribute.

  ### props.data

  **_Data Format_**

  The `data` data format is very similar to a two-dimensional array。

  - Each subarray in the array represents a row, and each element in the subarray represents a cell. The order in which cells are displayed is the order in which the arrays are indexed, so the order and display of each cell should be determined when reconstructing the data.

  ```
    [
      [cell, cell, cell], // row
      [
        cell, // cell
        cell,
        cell
      ],
      [cell, cell, cell],
      ...
    ]
  ```
  
  - An object can also represent a row in a two-dimensional array. This object has two required attributes: `type='row'` and `cells=[]`, which respectively identify the object as a row and the cells contained in the row:
  
  ```
  [
    {  // The rows are displayed as objects, and the in-row cells are contained in the cells field as an array.
      type: 'row',
      cells: [cell, cell, cell],
      ...
    },
    [  // Display rows as an array, each element representing a cell
      {type: 'button', ...},   // Generate a button in the cell
      cell,  // cell
      cell
    ],
    [
      cell,
      cell,
      [
          {type: 'radio'}, {type: 'radio'}, ... // Generate two or more radios in the same cell
      ]
    ],
    ...
  ]
  ```

  **_Object Unit_**

  The data format that cells can parse is divided into the following four categories:

  - String
  - Array，These four data formats can be nested inside the array again.
  - jsx,Such as: `<button className='test-button' onclick='()=>{return null}'>click</button>`
  - Object（Object cell），Details are as follows：

  |type                         | description                                                                                                                                               |
  |-----------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------|
  |row                          |Generate a unit row, `unit row object` can only be written in the body, invalid elsewhere                                                                  |
  |img                          |Generate an img tag inside the cell                                                                                                                        |
  |button                       |Generate a button in the cell                                                                                                                              |
  |link                         |Generate a link in the cell (a tag)                                                                                                                        |
  |radio                        |Generate a radio in the cell                                                                                                                               |
  |checkbox                     |Generate a checkbox in the cell                                                                                                                            |
  |select <sup>^1.4.1</sup>     |Generate a select in the cell                                                                                                                              |
  |text <sup>^1.5.0</sup>       |Generate a plain text inside the cell. You can also replace the object unit with a string. If you don't plan to add events to this cell, you can do that.  |
  |input <sup>^1.6.0</sup>      |Generate a text box in a cell.                                                                                                                             |

  > Here are some examples：
  >
  > ```
  > row
  >
  > {
  >   type: 'row',
  >   cells: [cell, cell, cell],
  >   data: 'row.id',
  >   value: 'row.typeID',
  >   event: 'onClick',
  >   callback: (data, cellData, event) => {},
  >   className: '',
  >   key: ''
  > }
  > 
  > button
  > 
  > {
  >   type: 'button',
  >   value: 'click me',
  >   className: 'test-btn',
  >   data: '123',
  >   event: 'onClick',
  >   callback: data => alert('hello react-tabllist', data) // hello react-tabllist, 123,
  >   key: ''
  > }
  > 
  > img
  > 
  > {
  >   type: 'img',
  >   src: 'http://www.xieyangogo.cn/pic.png',
  >   alt: '',
  >   text: 'IMG description',
  >   className: 'test-img',
  >   key: '',
  >   value:''
  > }
  > 
  > link (Choose one of the two)
  > 
  > {
  >   type: 'link',
  >   text: 'I am a link, I use the href attribute',
  >   className: 'test-link',
  >   key: '',
  >   href: 'https://github.com/oceanxy/react-tabllist',
  >   value:''
  > }
  > {
  >   type: 'link',
  >   text: 'I am a link, I use event and callback to implement custom functions',
  >   className: 'test-link',
  >   key: ''，
  >   data:  {},
  >   event: 'onClick',
  >   callback: (data, cellData, event) => {},
  >   value:''
  > }
  > 
  > radio
  > 
  > {
  >   type: 'radio',
  >   name: 'group2',
  >   text: 'radio group 2-1',
  >   className: 'test-radio',
  >   callback: (data, cellData, event) => {},
  >   key: '',
  >   value:''
  > }
  > 
  > checkbox
  > 
  > {
  >   type: 'checkbox',
  >   name: 'checkbox1',
  >   text: 'checkbox',
  >   className: 'test-checkbox',
  >   callback: (data, cellData, event) => {},
  >   key: '',
  >   value:''
  > }
  > 
  > select
  > 
  > {
  >   type: 'select',
  >   text: 'please choose：',
  >   value: '',
  >   data: '',
  >   className: '',
  >   callback: (data, cellData, event) => {},
  >   option: [
  >     {
  >       id: '1',
  >       label: 'item 1',
  >       value: 1
  >     }
  >   ]
  > }
  > 
  > text
  > 
  > {
  >   type: 'text',
  >   text: 'I am a normal text',
  >   callback: (data, cellData, event) => {}
  > }
  > 
  > input
  > 
  > {
  >   type: 'input',
  >   text: 'username:',
  >   placeholder: 'enter username',
  >   defaultValue: 'oceanxy' // This attribute is not a standard HTML tag attribute, but it is a property of the React element (the purpose is to set the default value of the text box). As you can see, you can also use React's officially defined properties in object cells.
  > }
  > ```

  **_Object Unit Attribute_**

  |**Attribute** `{type}`                                                 |Description                                                                                                                                                                                                                                                                          |row   |button  |link  |img       |radio     |checkbox  |input    |select   |text      |
  |-----------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------|--------|------|----------|----------|----------|---------|---------|----------|
  |~~**uid**~~ <br> `{string}` **Available before version 1.2.2**         |`Deprecated` A unique identifier for a cell that can be used to save ids, and so on. This field function is similar to the key, and can also be replaced with a data field in conjunction with a callback function, so it is decided to discard this field in version 1.2.2.             |❌    |✅     |✅    |✅       |✅        |✅        |❌      |❌       |❌        |
  |**type** <br> `{string}`                                               |The type of HTML tag to be generated in the cell                                                                                                                                                                                                                                     |✅    |✅     |✅    |✅       |✅        |✅        |✅      |✅       |✅        |
  |**text** <br> `{string}`                                               |The text of the rendered HTML tag                                                                                                                                                                                                                                                    |❌    |❌     |✅    |✅       |✅        |✅        |✅      |✅       |✅        |
  |**event** <br> `{string}`                                              |The way to trigger an event needs to be used with [callback](#Callback)                                                                                                                                                                                                              |✅    |✅     |✅    |❌       |✅        |✅        |✅      |✅       |✅        |
  |**callback** <br> `{function}`                                         |The callback function after the trigger event, see the [callback](#Callback) for details.                                                                                                                                                                                            |✅    |✅     |✅    |❌       |✅        |✅        |✅      |✅       |✅        |
  |**cells** <br> `{object[]}`                                            |Can only be used with the `row` type, see the [data](#props.data) for details.                                                                                                                                                                                                       |✅    |❌     |❌    |❌       |❌        |❌        |❌      |❌       |❌        |
  |**data** <br>`{*}`                                                     |Custom attributes can theoretically pass any value. This value is not used inside the component, you can get this value in the first parameter of the [callback](#Callback)                                                                                                           |✅    |✅     |✅    |✅       |✅        |✅        |✅      |✅       |✅        |
  |**option** <br> `{object[]}`                                           |Can only be used with the `select` type, see the [option](#Option) for details.                                                                                                                                                                                                      |❌    |❌     |❌    |❌       |❌        |❌        |❌      |✅       |❌        |


  **_HTML Attribute_**

  > The usage of these attributes is the same as that of native HTML tags. They are listed to indicate the points to be aware of when using them.

  |**Attribute** `{type}`                                                 |Description                                                                                                                                                                                           |
  |-----------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
  |**name** <br>`{string}`                                                |The attributes that the radio and checkbox must set, the same as the name attribute of the HTML-like tag.                                                                                             |
  |**value** <br> `{number\|string}`                                      |The `value` attribute of the generated HTML tag, `**All object cell can be used starting with version 1.4.1**`                                                                                        |
  |**src** <br> `{string}`                                                |Image link, such as: 'http(s)://xxx' or 'data:image/xxx'                                                                                                                                              |
  |**alt** <br> `{string}`                                                |The alt attribute of the image                                                                                                                                                                        |
  |**href** <br> `{string}`                                               |The hyperlink type of the link type (the same as the href of the HTML a tag), or you can use this combination of event and callback to customize the event callback without passing this attribute.   |

  > For more tag attributes please visit [w3schools](https://www.w3schools.com/tags/).

  **_React Element Attribute_**

  |**Attribute** `{type}`                                                 |Description                                                                                                                                                                                           |
  |-----------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
  |**key** <br> `{string}`                                                |Jsx loop or array need to use the key attribute, `Please ensure the uniqueness of the key`                                                                                                            |

CHANGELOG.md
# react-tabllist change log

---

#### 2.0.0[plan]
1. Add a panel for dynamically adjusting properties in the development environment.
2. New Function: A cell can bind multiple events.
3. Refactoring most of the old code.

#### 1.7.1[dev]
1. Modify the document according to the updated content of v1.7.0.
2. 修改bug-滚动到指定行的索引小于当前可视区域内第一行的索引时，滚动异常.

#### 1.7.0[latest]
1. Update dependency package.
2. Rewrite the scrolling logic; improve some comments, and modify some incorrect comments.
3. Improve the readability of some source code, such as variable naming.
4. Modify the className of the outermost container of the component to more recognizable and unique characters to avoid polluting other components or pages.
5. Cancel the mouse events of the outermost container and the events bound to the cell row in the code, and replace the original functions with React events. See Article 6 for details.
6. Added React event support, and 3 additional scroll-related custom events.
7. The disclosed attributes and methods have been optimized and adjusted.

#### 1.6.1
This version has not updated features, only some field names and comments have been optimized. In addition, the version of the core dependency library has been updated, as well as some updates in the development mode. The details are as follows:

1. Remove yarn.lock and package-lock.json from .gitignore
2. Update the version of highlight.js in development mode. The version used before the update (v 9.15.6) has a vulnerability. In fact, the version before v 10.4.1 has this vulnerability
3. Update the version of the core dependency library
4. Optimized the parameter structure of the callback function exposed to the object unit (note that props can be re-assigned through the callback function of the object unit, and the component has been updated. This function has existed before, but it is not explained in the text)
5. Update the README.md file

#### 1.6.0
1. New object cell: input.
2. Optimization of generation rules for object cells. You can now add custom attributes, react-related HTML attributes, and HTML attributes specified by w3school in the configuration of the object cell to make the resulting tags more friendly.
3. Modify the description of the fields available in the object cell in the README.md file.

#### 1.5.1 `For some reason, v1.5.1 has disappeared! Don't panic, everything in v1.5.1 is included in v1.6.0. But I recommend using the latest version.`
1. Fix bug: when setting css property {border-collapse: 'collapse'}, 'border-spacing' is invalid because it is ignored.
2. Fix bug: Numeric parsing error when custom cell width value is 'avg' or multiple values separated by commas.
3. Fix bug: When deactivating list auto scrolling, you can't manually control the list to scroll to the specified row.
4. Fix bug: Callback functions for object cell {type: 'radio', ...} and {type: 'checkbox', ...} are not available. Note: This bug exists in all versions prior to version 1.5.1.
5. Optimize the parameter structure of the callback function.
6. Refactored part of the code.

#### 1.5.0
1. Customize the text of the header serial number.
2. Configuration item update: property.body.row.rowCheckbox is no longer set directly to a Boolean value, but an object.
3. Custom row selection box or row serial number inserted into which column of the list.
4. You can define partial styles of the body, which usually do not affect the layout of the list.
5. Body height changes with list height whenever.
6. Fix bug: When the row selection or row number feature is enabled, the columns in the header are misaligned with the columns in the body.
7. Refactoring the code logic for row hover effects.
8. Normal text can also be written as an object cell to increase the event.

#### 1.4.3
1. Modify the logic of scrolling.
2. The height calculation method of the scroll area is updated: the padding of the component and the width of the border are now subtracted.
3. Performance optimization: Pause component scrolling when switching to other tabbed pages in the browser.

#### 1.4.2
1. Fix default parameters of property.scroll, consistent with previous versions.

#### 1.4.1
1. Fix bug: when scrolling by row and scrolling to the last row, the component will stop scrolling.
2. New object cell: drop-down list.
3. Some methods of the component can be used to manipulate the component in the object cell's callback function. For example: `scrollTo(9)` can scroll the list to a row with index 9.
4. update README document.

#### 1.4.0
1. Integrate the original scrolling function related components of the component and add some new scrolling configuration. You can configure it in ‘property.scroll’.

#### 1.3.1
1. Fix bug: When obsolete properties are used, the program is terminated when the obsolete property and the new substitution property (if any) do not belong to the child of the same object.
2. Print outdated property warnings only in the development environment.

#### 1.3.0
1. Added a description of modifying the data structure in the readme file: a new way of writing about row. (This feature is already available in version 1.2.1, but not in the readme file)
2. Fix bug: jsx is invalid in data.
3. Rewrite button, radio, and checkbox custom events; Rewrite the logic of radio and checkbox.
4. Readme file can be switched to Chinese or English.
5. A warning will be printed on the console when using the obsolete property.

#### 1.2.1
1. Cancel the definition of the click event of the row in the property in the 1.2.0 version, instead of defining it in the data property, just like other object cells.

#### 1.2.0
1. Events can be added to rows, and previous versions can only add events to cells.
2. Fix bug: When the line number feature is enabled, the style of the row number cell is applied to all cells in the row.

#### 1.1.4
1. Added two examples to be applied to actual projects.

app/index.js
import React from 'react'
import ReactDOM from 'react-dom'
import Tabllist from '../src'
import './index.scss'

const Dev = () => {
  const checkboxEvent = function(a, b, c) {
    console.log(a, b, c)
  }

  const option = {
    data: [
      [
        {
          type: 'input',
          text: '1',
          defaultValue: '1st column',
          placeholder: '1st column'
        },
        {
          type: 'text',
          text: '2nd column'
        },
        '3rd column',
        {
          type: 'select',
          text: '4th column',
          data: 123,
          className: '',
          option: [
            {
              id: '1',
              label: 'Scroll to the 2nd row',
              value: 0
            },
            {
              id: '2',
              label: 'Scroll to the 5rd row',
              value: 1
            },
            {
              id: '3',
              label: 'Scroll to the 7rd row',
              value: 2
            }
          ],
          event: 'onChange',
          callback: (instance, objectUnit, event) => {
            // step 1: Get the value of select
            const { value } = event.target
            // step 2: According to the value of select to match the value of the corresponding row in the data,
            // 				 and then get the index of the row
            const { scrollTo, renderData } = instance
            for (let i = 0, k = renderData; i < k.length; i++) {
              // eslint-disable-next-line no-undef
              if (_.isPlainObject(renderData[i]) && parseInt(renderData[i].value) === parseInt(value)) {
                // step 3: Call method scrolling list
                scrollTo(i - 1)
                break
              }
            }
          }
        }
      ],
      [
        <span key={0}>I am a span</span>,
        <div key={1} onClick={() => alert('test JSX event')}>test JSX event</div>,
        <a key={2} href="http://www.xieyangogo.cn/react-tabllist/">I am a link</a>,
        <div key={3}>I am a div</div>
      ],
      {
        type: 'row',
        data: 1,
        value: 0,
        event: 'onClick',
        callback: (restData, objectUnit, event) => {
          alert('test event of row')
          console.log(restData, objectUnit, event)
        },
        className: 'click-row',
        cells: [
          'row 1; column 1',
          {
            type: 'link',
            text: 'I am a first link',
            className: 'test-link',
            callback: () => {
              console.log('I am a first link')
            }
          },
          {
            type: 'link',
            text: 'I am a second link',
            href: 'https://github.com/oceanxy/react-tabllist',
            className: 'test-link'
          },
          {
            type: 'button',
            value: 'click me',
            className: 'test-btn',
            callback: () => {
              alert('hello react-tabllist')
            }
          }
        ]
      },
      [
        {
          type: 'img',
          src: 'http://www.xieyangogo.cn/root/assets/img/web_github_128px.png',
          alt: '',
          text: 'row 2; column 1',
          key: '',
          value: '',
          style: {
            width: 24,
            height: 24
          }
        },
        {
          type: 'link',
          data: 10,
          text: 'I am a link, I use event and callback to implement custom functions',
          className: 'test-link2',
          key: '',
          event: 'onClick',
          callback: () => {
            window.open('https://github.com/oceanxy/react-tabllist')
          }
        },
        {
          type: 'link',
          text: 'I am a link',
          href: 'https://github.com/oceanxy/react-tabllist',
          className: 'test-link'
        },
        {
          type: 'button',
          value: 'click me',
          className: 'test-btn',
          event: 'onClick',
          callback: (data, cellObject, obj) => {
            obj.target.value = 'clicked'
          }
        }
      ],
      ['row 3; column 1', 'row 3; column 2', 'row 3; column 3', 'row 3; column 4'],
      {
        type: 'row',
        data: 1,
        value: 1,
        event: 'onClick',
        callback: (restData, objectUnit, event) => {
          alert('test event of row')
          console.log(restData, objectUnit, event)
        },
        className: 'click-row',
        cells: [
          {
            type: 'link',
            text: 'I am a link',
            href: 'https://github.com/oceanxy/react-tabllist',
            className: 'test-link'
          },
          {
            type: 'button',
            value: 'click me',
            className: 'test-btn',
            callback: () => {
              alert('hello react-tabllist')
            }
          },
          'row 4; column 3',
          'row 4; column 4'
        ]
      },
      [
        [
          {
            type: 'radio',
            name: 'group1',
            text: 'row 5; column 1-1',
            value: 0,
            className: 'test-radio',
            callback: checkboxEvent
          },
          {
            type: 'radio',
            name: 'group1',
            text: 'row 5; column 1-2',
            value: 1,
            className: 'test-radio',
            callback: checkboxEvent
          }
        ],
        'row 5; column 2',
        'row 5; column 3',
        'row 5; column 4'
      ],
      {
        type: 'row',
        data: 1,
        value: 2,
        event: 'onClick',
        callback: (restData, objectUnit, event) => {
          alert('test event of row')
          console.log(restData, objectUnit, event)
        },
        className: 'click-row',
        cells: [
          [
            {
              type: 'radio',
              name: 'group2',
              text: 'radio group 2-1',
              className: 'test-radio',
              value: 'a',
              callback: checkboxEvent
            },
            {
              type: 'radio',
              name: 'group2',
              text: 'radio group 2-2',
              className: 'test-radio',
              value: 'b',
              callback: checkboxEvent
            }
          ],
          'row 6; column 2',
          'row 6; column 3',
          'row 6; column 4'
        ]
      },
      [
        [
          {
            type: 'checkbox',
            name: 'chk',
            text: 'chk1',
            value: 'chk1',
            callback: checkboxEvent
          },
          {
            type: 'checkbox',
            name: 'chk',
            text: 'chk2',
            value: 'chk2',
            callback: checkboxEvent
          },
          {
            type: 'checkbox',
            name: 'chk',
            text: 'chk3',
            value: 'chk3',
            callback: checkboxEvent
          }
        ],
        {
          type: 'link',
          text: 'I am a link',
          event: 'onClick',
          className: 'test-link',
          callback: () => {
            alert('clicked link')
          }
        },
        {
          type: 'button',
          value: 'click me',
          className: 'test-btn',
          callback: (data, cellObject, cellElement) => {
            if (!data) {
              data = 'data of button is undefined'
            }

            cellElement.target.value = 'you clicked me!!'
            cellElement.target.style.width = '150px'

            console.log(data)
            console.log(cellObject)
            console.log(cellElement)

            alert('hello react-tabllist, Please check the console')
          }
        }
      ],
      ['row 8; column 1', 'row 8; column 2', 'row 8; column 3', 'row 8; column 4'],
      ['row 9; column 1', 'row 9; column 2', 'row 9; column 3', 'row 9; column 4'],
      ['row 10; column 1', 'row 10; column 2', 'row 10; column 3', 'row 10; column 4']
    ],
    property: {
      border: {
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: '#ededed'
      },
      style: {
        width: '100%',
        margin: '0 auto',
        padding: '10px',
        height: 550,
        border: '1px solid #999999'
      },
      scroll: {
        enable: true,
        // speed: 1000,
        // distance: -1
        speed: 1000,
        distance: -2
      },
      header: {
        show: true,
        style: {
          backgroundColor: '#1693ff',
          height: 40,
          borderBottom: '1px solid #999999'
        },
        cellStyle: {
          color: '#ffffff',
          fontWeight: 'bolder',
          fontSize: 20
        }
      },
      body: {
        style: {
          padding: 0
        },
        row: {
          spacing: 0,
          style: {
            height: 34
          },
          rowCheckbox: {
            show: true,
            column: 1
          },
          serialNumber: {
            show: true,
            formatter: 'No.{index}',
            column: 1,
            style: {
              backgroundColor: '#3991ff',
              width: 80,
              fontSize: 20,
              color: '#2cff41',
              margin: '0 auto'
            },
            specialStyle: [
              { backgroundColor: '#203d7b' },
              { backgroundColor: '#2f4c99' },
              { backgroundColor: '#3960c0' }
            ]
          },
          specialStyle: [
            { height: 60 },
            { height: 100 },
            { height: 80 },
            { height: 30 },
            { height: 50 },
            { height: 80 },
            {},
            { height: 200 }
          ],
          visual: {
            show: true,
            style: {
              backgroundColor: '#e8f4fc'
            }
          },
          silent: {
            show: false,
            style: {
              backgroundColor: '#bcf0fc'
            }
          }
        },
        cell: {
          style: {
            fontSize: 16,
            minWidth: 0,
            color: '#000000',
            textAlign: 'center',
            width: '80, 50, 400, ,150'
          }
        },
        cellOfColumn: {
          style: [
            {},
            {},
            { color: 'red' },
            {}
          ]
        }
      }
    }
  }

  return (
    <div className="container">
      <Tabllist
        property={option.property}
        data={option.data}
        onMouseOver={(e, instance) => {
          instance.pause(true)
        }}
        onMouseOut={(e, instance) => instance.pause(false)}
        // onScroll={
        //   () => {
        //     // some code
        //   }
        // }
        // custom_onScrollTo={
        //   // 一次性滚动（scroll.distance<0时）结束后或调用公开的scrollTo函数后触发
        //   (e, instance) => console.log(e, instance)
        // }
        // custom_onScrollToEnd={
        //   // 当distance>=0时，滚动到组件尾部触发，即组件最后一行完全进入可视区域内时触发
        //   (e, instance) => console.log(e, instance)
        // }
        // custom_onScrollFullCircle={
        //   // 滚动完整一圈后触发
        //   (e, instance) => console.log(e, instance)
        // }
      />
    </div>
  )
}

ReactDOM.render(
  <Dev />,
  document.getElementById('root')
)
